h2o-core/src/main/java/water/parser/FVecParseWriter.java

```diff
@@ -157,19 +172,37 @@ public FVecParseWriter(Vec.VectorGroup vg, int cidx, Categorical[] categoricals,
   @Override public final void rollbackLine() {}
 
   @Override public void invalidLine(ParseErr err) {
-    addErr(err);
+    addError(err);
     newLine();
   }
 
   @Override
-  public void addError(ParseErr err) {
+  public final void addError(ParseErr err) {
     if(_errs == null)
       _errs = new ParseErr[]{err};
-    else  if(_errs.length < 20)
+    else if(_errs.length < MAX_ERR_CNT)
       _errs = ArrayUtils.append(_errs,err);
     _errCnt++;
   }
 
+  private void addMissingColumnsError() {
+    assert _maxMissingCol >= _nCols;
+    String message = "Invalid line, found more columns than expected (found: " + (_maxMissingCol + 1) + ", expected: " + _nCols + ")" +
+            ((_missingColVals == null) ? "" : "; values = {" + toStringTrimmedCleaned(_missingColVals, 50) + "}");
+    addError(new ParseErr(message, _cidx, lineNum()));
+  }
+
+  private static String toStringTrimmedCleaned(StringBuilder sb, int maxLen) {
+    String msg = sb.length() <= maxLen ? sb.toString() : sb.substring(0, maxLen) + "...(truncated)";
+    return msg.replaceAll("[^\\x00-\\x7F]", ""); // Replace non-ASCII characters to avoid problems with displaying of the message in client
+  }
+
+  private void recordMissingColumnError(int colIdx, Object value) {
+    if ((_maxMissingCol == -1) || (_maxMissingCol + 1 == colIdx)) // we want to report values in the right order (or not at all)
+      _missingColVals = (_missingColVals == null ? new StringBuilder() : _missingColVals.append(", ")).append(value);
```

Comment:
I think we should put a max count on how many missing columns we report, just in case the data is broken and there is a multi megabyte line.